Movie Search/Services/NetworkAPI.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  NetworkAPI.swift
+//  Movie Search
+//
+//  Created by Alexey Grabik on 15/04/2020.
+//
+
+import UIKit
+
+protocol NetworkService {
+    func search(for movie: String, completionHandler: @escaping ([MovieModel]?, NetworkServiceError?) -> Void)
+    func getPoster(_ poster: String, completionHandler: @escaping (UIImage?) -> Void) -> Disposable?
+}
+
+enum NetworkServiceError {
+    case badResponse, parsingError, constructingRequestError
+}
+
+class Disposable {
+    private let cancelHandler: () -> Void
+    
+    init(_ cancelHandler: @escaping () -> Void) {
+        self.cancelHandler = cancelHandler
+    }
+    
+    func dispose() {
+        cancelHandler()
+    }
+}
+
+class NetworkAPI: NSObject, NetworkService {
+    
+    private let imageCache = ImageCache()
+    lazy private var session = URLSession.shared
+    
+    typealias ImageRequestHandler = (UIImage?) -> Void
+    private var imageRequestsHandlers = [URL: [ImageRequestHandler]]()
+    
+    func search(for movie: String, completionHandler: @escaping ([MovieModel]?, NetworkServiceError?) -> Void) {
+        var urlComponents = URLComponents(string: Constants.searchAPIURLString)
+        urlComponents?.queryItems = [ URLQueryItem(name: "api_key", value: Constants.APIKey),
+                                      URLQueryItem(name: "query", value: movie) ]
+
+        guard let url = urlComponents?.url else {
+            completionHandler(nil, NetworkServiceError.constructingRequestError)
+            return
+        }
+        
+        query(url) { data, error in
+            guard let data = data else {
+                DispatchQueue.main.async {
+                    completionHandler(nil, NetworkServiceError.badResponse)
+                }
+                return
+            }
+            do {
+                let page = try JSONDecoder().decode(PageModel.self, from: data)
+                let moviesModels = page.results
+                DispatchQueue.main.async {
+                    completionHandler(moviesModels, nil)
+                }
+            }
+            catch {
+                DispatchQueue.main.async {
+                    completionHandler(nil, NetworkServiceError.parsingError)
+                }
+            }
+        }
+    }
+    
+    func getPoster(_ poster: String, completionHandler: @escaping (UIImage?) -> Void) -> Disposable? {
+        assert(Thread.isMainThread)
```

Comment:
I noticed this in a few different places - why is it needed?

Reply:
to ensure that there is only one thread accessing imageRequestsHandlers at a time